sdk/azcore/policy/policy.go

```diff
@@ -132,5 +133,28 @@ type TokenRequestOptions struct {
 
 // BearerTokenOptions configures the bearer token policy's behavior.
 type BearerTokenOptions struct {
-	// placeholder for future options
+	// AuthorizationHandler allows SDK developers to run client-specific logic when BearerTokenPolicy must authorize a request.
+	// When this field isn't set, the policy follows its default behavior of authorizing every request with a bearer token from
+	// its given credential.
+	AuthorizationHandler AuthorizationHandler
+}
+
+// AuthorizationHandler allows SDK developers to insert custom logic that runs when BearerTokenPolicy must authorize a request.
+// Errors returned by its functions should be have a NonRetriable() marker method to prevent [runtime.RetryPolicy] retrying
```

Comment:
Any reason to not make this a formal requirement?
```suggestion
// Errors returned must have a NonRetriable() marker method to prevent [runtime.RetryPolicy] retrying
```

Reply:
If we always want this to return a non-retriable error, we should just check for `NonRetriable()` and if it's not there then wrap it ourselves.

Reply:
I'll implement the wrapping approach. There may be a scenario in which retrying after an error from one of these functions is the right thing to do, but that would be an unusual scenario (let me know if you think of one; I haven't). It's more likely someone will trigger pointless retries by returning a plain old error from one of these functions.